File: src/frontend/public-website/src/api/rankings.ts
import { apiCall, buildQueryString } from './client';
import type { ModalityRanking, GeneralRankingResponse } from './types';

export const rankingsApi = {
  /**
   * Get general ranking (all courses)
   */
  getGeneralRanking: async (seasonId?: string): Promise<GeneralRankingResponse> => {
    const query = seasonId ? buildQueryString({ season_id: seasonId }) : '';
    return apiCall<GeneralRankingResponse>(`/rankings/general${query}`);
  },

  /**
   * Get modality ranking
   */
  getModalityRanking: async (modalityId: string, seasonId?: string): Promise<ModalityRanking> => {
    const query = seasonId ? buildQueryString({ season_id: seasonId }) : '';
    return apiCall<ModalityRanking>(`/rankings/modality/${modalityId}${query}`);
  },

  /**
   * Get course ranking (breakdown by modality)
   */
  getCourseRanking: async (courseId: string, seasonId?: string): Promise<{ [modalityName: string]: number }> => {
    const query = seasonId ? buildQueryString({ season_id: seasonId }) : '';
    return apiCall<{ [modalityName: string]: number }>(`/rankings/course/${courseId}${query}`);
  },
};

File: src/apis/public-api/app/schemas/rankings.py
"""
Rankings-related schemas
"""

from typing import List

from pydantic import BaseModel


class RankingEntry(BaseModel):
    """Model for a ranking entry"""

    position: int
    course_id: int
    course_name: str
    course_short_code: str
    points: int
    played: int
    won: int
    drawn: int
    lost: int


class ModalityRanking(BaseModel):
    """Model for modality-specific ranking"""

    modality_id: int
    modality_name: str
    rankings: List[RankingEntry]


class CourseRanking(BaseModel):
    """Model for course-specific ranking"""

    course_id: int
    course_name: str
    modalities: List[ModalityRanking]
    total_points: int


class GeneralRanking(BaseModel):
    """Model for general ranking"""

    season_id: int
    season_year: int
    rankings: List[RankingEntry]
